src/canister_tests/src/certificate_validation.rs

```diff
@@ -0,0 +1,166 @@
+    }
+
+    // 3. (cont.) The subnet state tree in the certificate must reveal the canister's certified data.
+    let certified_data_path = vec![
+        "canister".into(),
+        canister_id.into(),
+        "certified_data".into(),
+    ];
+    let witness = match lookup_value(&cert, certified_data_path) {
+        Ok(witness) => witness,
+        Err(err) => {
+            return Err(WitnessLookupFailed { inner: err });
+        }
+    };
+
+    // 4. The tree must be a hash tree as per Encoding of certificates.
+    let tree: HashTree = decode_value(encoded_tree)?;
+    // 5. The root hash of that tree must match the canister's certified data.
+    if *witness != tree.digest() {
+        return Err(TreeHashCertifiedDataMismatch);
+    }
+
+    // 6. The path ["http_assets",<url>], where url is the utf8-encoded url from the HttpRequest must exist and be a leaf.
+    // Else, if it does not exist, ["http_assets","/index.html"] must exist and be a leaf.
+    let asset_path = ["http_assets".into(), uri_path.into()];
+    let tree_sha = match tree.lookup_path(&asset_path) {
+        LookupResult::Found(v) => v,
+        _ => match tree.lookup_path(&["http_assets".into(), "/index.html".into()]) {
+            LookupResult::Found(v) => v,
+            _ => {
+                return Err(AssetPathLookupFailed);
+            }
+        },
+    };
+
+    // 7. That leaf must contain the SHA-256 hash of the decoded body.
+    // This is where Internet Identity breaks spec because it certifies encoded response bodies, see L2-722 for details.
+    let body_sha = decode_body_to_sha256(body, encoding).unwrap();
+    if body_sha != tree_sha {
+        return Err(AssetHashMismatch);
+    }
+    Ok(())
+}
+
+fn parse_header(ic_certificate: &str) -> Result<(&str, &str), ValidationError> {
+    let captures = Regex::new("^certificate=:([^:]*):,\\s*tree=:([^:]*):$")
+        .unwrap()
+        .captures(ic_certificate)
+        .ok_or(MalformedCertificate {
+            message: "unexpected format".to_string(),
+        })?;
+    let encoded_cert = captures
+        .get(1)
+        .ok_or(MalformedCertificate {
+            message: "no match for encoded cert".to_string(),
+        })?
+        .as_str();
+    let encoded_tree = captures
+        .get(2)
+        .ok_or(MalformedCertificate {
+            message: "no match for encoded tree".to_string(),
+        })?
+        .as_str();
+    Ok((encoded_cert, encoded_tree))
+}
+
+fn agent(root_key: &ThresholdSigPublicKey) -> Agent {
+    let agent = Agent::builder()
+        .with_transport(
+            ReqwestHttpReplicaV2Transport::create("https://irrelevant-but-mandatory.com").unwrap(),
+        )
+        .build()
+        .unwrap();
+    agent
+        .set_root_key(public_key_to_der(&root_key.into_bytes()).unwrap())
+        .expect("setting root key failed");
+    agent
+}
+
+fn decode_value<T>(encoded_value: &str) -> Result<T, ValidationError>
```

Comment:
nit: for once I wouldn't mind a name that's a little longer; `decode_value` could be decoding anything. Or maybe something like `decode_cert`?